Introduce resource oriented API (#395)
ietf-wg-ppm/draft-ietf-ppm-dap: draft-ietf-ppm-dap.md

# Protocol Definition

DAP is structured as an HTTP application consisting of three major interactions
or sub-protocols:

* Uploading reports from the client to the aggregators, using the HPKE
  configuration ({{hpke-config-resource}}) and report ({{report-resource}})
  resources, as described in {{upload-flow}}.

> * Computing the aggregate over sets of reports, using the aggregation job

Author: I think this would be a bit confusing, because in this bullet, we are discussing the aggregation sub-protocol, which works in terms of aggregation jobs, and those aren't necessarily equivalent to batches.

  resource ({{aggregation-job-resource}}), as described in {{aggregate-flow}}.
* Collecting aggregation results, using the aggregate share

>   ({{aggregate-share-resource}}) and collection ({{collection-resource}})

Author: I feel like that's trying to sneak a verb ("job") into a resource, which should be a noun. However this is purely a naming quibble so I can live with either. I'm curious what others think; I'd be happy to go with the majority consensus.

>   ({{aggregate-share-resource}}) and collection ({{collection-resource}})

Author: I put it back to "collect job" in the most recent round of changes. If you both like it better, we can change it again to "collection job".

>   ({{aggregate-share-resource}}) and collection ({{collection-resource}})

Reviewer: I'd be happy with either "collect job" or "collection job". (I have a slight preference for "collection job" as it matches "aggregation job" & is slightly more grammatically correct, but this is very quibbly.)

  resources, as described in {{collect-flow}}.

In this section, we discuss each of the resources and the messages used to act
on these resources. A resource's path is resolved relative to a server's
endpoint to construct a resource URI. A server's endpoint MAY contain
arbitrarily many path components before the DAP resources.

Resource paths are specified as templates like:

~~~
/resource_type/{resource-id}
~~~

> Templates may also include optional query parameters, indicated by square

Author: I don't know, it depends on what query params we add in the future. For now I'm trying to make the path templates no more complicated than what is needed to describe the API we currently have.

brackets ("[" and "]"), as in:

~~~
/resource_type/{resource-id}[?query_parameter={resource-id}]
~~~

DAP resource identifiers are always opaque byte strings, so anywhere
`{resource-id}` occurs in a template (e.g., `{task-id}` or `{report-id}`) MUST
be expanded to the URL-safe, unpadded Base 64 representation of the resource's

> identifier, as specified in sections 5 and 3.2 of {{!RFC4648}}.

Author: [RFC 6570](https://datatracker.ietf.org/doc/rfc6570/) discusses URI templating, and I considered using it here, but decided not to because (1) we don't need a template scheme as complicated as the one in 6570 and (2) 6570 has rules about how variables in templates should be percent encoded to be URL safe, whereas we have already settled on using URL-safe unpadded base64.

## Uploading Reports {#upload-flow}

Clients periodically upload reports to the Leader, which then distributes the
individual shares to each Helper. The upload sub-protocol involves the HPKE
configurations ({{hpke-config-resource}}) and reports ({{report-resource}})
resources.

### HPKE Configuration Resource {#hpke-config-resource}

Before the client can upload its report to the leader, it must know the HPKE
configuration of each aggregator. See {{compliance}} for information on HPKE
algorithm choices. Clients retrieve the HPKE configuration from each aggregator
by accessing the HPKE configuration resource.

#### Required methods

> ##### GET `/hpke_config[?task_id={task-id}]`

Reviewer 2: For consistency with other resources? Also, the return value is a list of HPKE configs.
```suggestion
##### GET `/hpke_configs[?task_id={task-id}]`
```

Author: Yes, excellent idea.

[OPEN ISSUE: Decide whether to expand the width of the id.]

The `HpkeConfigList` structure contains one or more `HpkeConfig` structures in
decreasing order of preference. This allows a server to support multiple HPKE
configurations simultaneously.

[TODO: Allow aggregators to return HTTP status code 403 Forbidden in deployments

> that use authentication to avoid leaking information about which tasks exist.]

Author: No, I guess not, since we don't have any text that forbids this.

} Report;
~~~
##### PUT `/tasks/{task-id}/reports/{report-id}`

`report-id` is a `ReportId` used by the Aggregators to ensure the report appears
in at most one batch (see {{input-share-validation}}). The Client MUST generate
this using a cryptographically secure random number generator.

This message is called the Client's report. It consists of the task ID, report
metadata, the "public share" output by the VDAF's input-distribution algorithm,
and the encrypted input share of each of the Aggregators. (Note that the public
share might be empty, depending on the VDAF. For example, Prio3 has an empty
public share, but Poplar1 does not. See {{!VDAF}}.) The header consists of the
task ID and report "metadata". The metadata consists of the following fields:

* A report ID used by the Aggregators to ensure the report appears in at most
  one batch. (See {{input-share-validation}}.) The Client MUST generate this by
  generating 16 random bytes using a cryptographically secure random number
  generator.

* A timestamp representing the time at which the report was generated.
  Specifically, the `time` field is set to the number of seconds elapsed since

>   the start of the UNIX epoch. The client SHOULD round this value down to the

Author: I moved it down to line 928, in the discussion of `struct Report`:

~~~
* `time` is the time at which the report was generated. The Client SHOULD round
  this value down to the nearest multiple of the task's `time_precision` in
  order to ensure that that the timestamp cannot be used to link a report back
  to the Client that generated it.
~~~

## Verifying and Aggregating Reports {#aggregate-flow}

Once a set of Clients have uploaded their reports to the Leader, the Leader can
send them to the Helpers to be verified and aggregated. Verification of a set of
reports is referred to as an aggregation job. Each aggregation job is associated
with exactly one DAP task, and a DAP task will have many aggregation jobs. In
order to enable the system to handle very large batches of reports, this process
can be parallelized by dividing the preparation of a batch's reports into
multiple aggregation jobs.

To run an aggregation job, the Leader sends a message to each Helper containing
the report shares in the job. The Helper then processes them (verifying the
proofs and incorporating the output shares into the ongoing aggregate) and
replies to the Leader.

The exact structure of the aggregation job flow depends on the VDAF.
Specifically:

* Some VDAFs (e.g., Prio3) allow the Leader to start aggregating reports
  proactively before all the reports in a batch are received. Others (e.g.,

>   Poplar1) cannot be prepared until the collector provides the aggregation

Reviewer 2: ```suggestion
  Poplar1) cannot be prepared until the Collector provides the aggregation
```

{: #aggregation-flow-illustration title="Aggregation Flow (batch size=20).
Multiple aggregation flows can be executed at the same time."}

The aggregation flow can be thought of as having two phases for transforming
each valid input report share into an output share:

- Initialization: Begin the aggregation flow by sharing report shares with each
  Helper. Each aggregator, including the Leader, initializes the underlying VDAF
  instance using these report shares and the VDAF configured for the
  corresponding measurement task.
- Continuation: Continue the aggregation flow by exchanging messages produced by
  the underlying VDAF instance until aggregation completes or an error occurs.
  These messages do not replay the shares.

### Aggregation Job Resource {#aggregation-job-resource}

Executing the aggregation flow consists of the Leader interacting with the
Helper's aggregate job resource. Only the Helper supports this resource.

#### Representation {#aggregation-job-representation}

~~~
enum {

>   continued(0),

Author: As noted [here](https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/pull/395#discussion_r1067551046), there should not be a start state.

  finished(1),
  failed(2),
  (255),
} PrepareStepState;

struct {
  ReportID report_id;
  PrepareStepState prepare_step_state;
  select (PrepareStep.prepare_step_state) {

>     case start: ReportShare;

Author: This is an error, left over from a previous iteration where a single `AggregationJob` message was used for all stages of the aggregate sub-protocol. There shouldn't be a `start` state here. Thank you for spotting that.

Multiple aggregation flows can be executed at the same time."}

The aggregation flow can be thought of as having three phases for transforming
each valid input report share into an output share:

- Initialization: Begin the aggregation flow by sharing report shares with each
  Helper. Each aggregator, including the Leader, initializes the underlying VDAF
  instance using these report shares and the VDAF configured for the
  corresponding measurement task.
- Continuation: Continue the aggregation flow by exchanging messages produced by
  the underlying VDAF instance until aggregation completes or an error occurs.
  These messages do not replay the shares.
- Completion: Finish the aggregate flow, yielding an output share corresponding
  to each input report share in the batch.

### Aggregate Initialization {#agg-init}

The Leader begins an aggregation job by choosing a set of candidate reports that
pertain to the same DAP task. The Leader can prepare many reports in parallel by
concurrently creating multiple aggregation jobs. After choosing the set of
candidates, the Leader begins aggregation by splitting each report into "report

> shares", one for each aggregator. The Leader and Helpers then run the aggregate

Reviewer 2: ```suggestion
shares", one for each Aggregator. The Leader and Helpers then run the aggregate
```

[[OPEN ISSUE: consider moving the helper report ID check into
{{input-share-validation}}]]

Once the Helper has processed each valid report share in
`AggregationJob.prepare_steps`, it responds with an `AggregationJob` (see
{{aggregation-job-representation}}). The order of its `prepare_steps` MUST match
the Leader's request. Each report that was marked as invalid is assigned the
`PrepareStepState` `failed`. Otherwise, the `PrepareStepState` is either
`continued` with the output `prep_msg`, or `finished` if the VDAF preparation

> process is finished for the report share.

Author: As explained [here](https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/pull/395#discussion_r1065098317), we're going to change this.

The Helper SHOULD store each report share's current preparation state and round.

Upon receipt of a Helper's `AggregationJob` message, the Leader checks that the
sequence of `PrepareStep` messages corresponds to the `PrepareStep` sequence of
the request. If any `PrepareStep` appears out of order, is missing, has an
unrecognized report ID, is in the `start` state, or if two prepare steps have
the same report ID, then the Leader MUST abort with error `unrecognizedMessage`.
If `AggregationJob.round` is not 1, then the Leader MUST abort with error

> `roundMismatch`.

Author: I don't think that can work: in the case of the helper replying to the leader, there's no request path in which to indicate the round.

> `roundMismatch`.

Reviewer 2: Discussed offline: This is just spelling.

`AggregationJob` (see {{aggregation-job-representation}}). The `round` field
MUST be 0. The order of the `prepare_steps` MUST match that of the
`report_shares` in the Leader's request (ordered by either structure's
`report_id` field). Each report that was marked as invalid is assigned the
`PrepareStepState` `failed`. Otherwise, the `PrepareStepState` is either
`continued` with the output `prep_msg`, or `finished` if the VDAF preparation
process is finished for the report share.

The Helper SHOULD store each report share's current preparation state and the
aggregation job's round.

##### POST `/tasks/{task-id}/aggregation_jobs/{aggregation-job-id}` {#aggregation-job-post}

In the continuation phase, the Leader drives the VDAF preparation of each share
in the candidate report set until the underlying VDAF moves into a terminal
state, yielding an output share for all aggregators or an error. This phase may
involve multiple rounds of interaction depending on the underlying VDAF. Each
round trip is initiated by the Leader making a POST request to an aggregation
job resource.

###### Leader Continuation {#aggregation-leader-continuation}

> The Leader begins each round of continuation with an `AggregationJob` received

Author: You're quite right that a leader could validate the helper's `AggregationJob` immediately. The reason I put this text here is so that I don't have to duplicate it at the bottom of both "Leader Initialization" and also "Leader Continuation" (in both cases, the Leader gets an `AggregationJob` from the helper and must validate it). But I think what we could do is add a section "Aggregation Job Validation" as a peer to "Input Share Preparation" and "Input Share Validation" and then refer to that from both "Leader Initialization" and "Leader Continuation".

- [ ] refactor aggregation job validation text into another section

finished, then the report cannot be processed further and MUST be abandoned.

If the Helper's `PrepareStep` is of type `continued`, then the Leader proceeds
as follows.

Let `leader_outbound` denote the Leader's prepare message and `helper_outbound`
denote the Helper's. The Leader advances its aggregation job to the next round
by computes the next state transition as follows:

~~~
inbound = VDAF.prep_shares_to_prep(agg_param, [leader_outbound, helper_outbound])
out = VDAF.prep_next(prep_state, inbound)
~~~

where `[leader_outbound, helper_outbound]` is a vector of two elements. If
either of these operations fails, then the Leader marks the report as invalid.
Otherwise, if the VDAF has reached the finished state, then `out` will be the
Leader's output share, in which case the Leader finishes and stores its output
share for further processing as described in {{collect-flow}}. Otherwise, `out`
is the pair `(new_state, prep_msg)`, where `new_state` is the Leader's updated
state and `prep_msg` is the next VDAF message (which will be `leader_outbound`
in the next round of continuation). For the latter case, the Leader sets
`prep_state` to `new_state`.

> The Leader SHOULD store each report share's current preparation state and the

Reviewer: Shouldn't this storage happen after receiving the response from the helper?

Author: I think there's a few different points at which the leader could store different things. What we care about is that the leader checkpoint state so that it can recover from crashes, but the more I think about it, the more I think it's inappropriate for the spec to say anything about what an implementation stores or when. So maybe we shouldn't say anything about this at all.

Author: We have the same problem in the equivalent helper text (noted by Brandon [here](https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/pull/395#discussion_r1067587961). These two passages are awkward, because there's a few different ways for implementations to checkpoint state such that they can implement round skew recovery, and I don't think it's appropriate for the spec to dictate implementation details (because maybe there's some nuance about a particular implementation's storage system that makes the prescribed strategy difficult or impossible). Maybe these SHOULDs about storing state should be deleted and instead we could have an implementation note elsewhere (Operational Considerations?) discussing the intent of the `round` field. and suggesting one possible solution.

request is an `AggregationJob` (see {{aggregation-job-representation}}).
`AggregationJob.round` MUST be set to the Leader's current round (i.e., one more
than the `round` value in the Helper's latest `AggregationJob` message) and the
`prepare_steps` field MUST be a sequence of `PrepareStep`s in the `continued`
state containing the corresponding `inbound` prepare message.

###### Helper Continuation

If the Helper does not recognize the task ID, then it MUST abort with error
`unrecognizedTask`. If the Helper does not recognize the aggregation job, then
it MUST abort with error `unrecognizedAggregationJob`. This could occur because
the aggregation job never existed, but also if it was deleted (see
{{aggregation-job-delete}}).

Otherwise, for each `PrepareStep` received from the Leader in
`AggregationJob.prepare_steps`, the Helper checks
`PrepareStep.prepare_step_state` to determine if the report share should
continue being prepared. If the report share is in state `failed` or `finished`,
then the Helper marks the report as failed with error `unrecognized_message`.

If the state is `continued`, then the Helper continues with preparation for the

> report share. The Helper MUST check its current preparation state against the

Reviewer: ```suggestion
report share. The Helper MUST check its current round against the
```

Leader's `AggregationJob.round` value. If the Leader is one round ahead of the
Helper, then the Helper can combine the Leader's prepare message and the
Helper's current preparation state (`prep_state`) as follows:

~~~
out = VDAF.prep_next(prep_state, inbound)
~~~

where `inbound` is the previous VDAF prepare message sent by the Leader and
`prep_state` is the Helper's current preparation state. This step yields one of
three outputs:

1. An error, in which case the input report share is marked as failed and the
   Helper replies to the Leader with a `PrepareStep` in the `failed` state.
1. An output share, in which case the Helper stores the output share for future
   collection as described in {{collect-flow}} and replies to the Leader with a
   `PrepareStep` in the `finished` state. The Helper expects no further requests
   from the Leader for a report in the `finished` state.
1. An updated VDAF state and preparation message, denoted
   `(prep_state, prep_msg)`, in which case the Helper replies to the leader with
   `PrepareStep` in the `continued` state containing `prep_msg`.

> After computing each report share's preparation state and prepare message, the

Reviewer: Recovering from round skew only requires storing one old round of responses; I think it would be worthwhile to mention this.

Author: - [ ] include implementation note explaining that helper only needs to store the most recent set of prepare messages for each aggregation job

> Helper SHOULD store them, tagged with the round they were computed in, so that

Reviewer: Here we say the helper should store the answer, but below we say the helper can either store or recompute the response (paragraph starting L1404).

I'd suggest dropping the "recompute" part -- storing the answer is simpler & less error-prone. And recomputing the response still requires storing information to allow the recomputation to occur, I think.

Author: I think this is similar to the equivalent leader text discussed [here](https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/pull/395#discussion_r1067581728) so I'll pursue discussion there.

Reviewer: The only part of this thread that's different from the linked thread is that this thread suggests dropping the suggestion to "recompute" -- IIUC, recomputing is a strictly inferior strategy to storing the previously-computed result (since even in a recompute strategy, you still need to store the data required to recompute the result, i.e. recomputation also requires storage).

Agreed w/ your comment in the other thread that we should probably not be very specific about when storage occurs.

Author: I deleted the SHOULDs about storing anything and punted the discussion of round skew to a new "Recovering From Round Skew".

the Leader can recover from losing the Helper's response.

If the `round` in the Leader's request is equal to the Helper's current round
(i.e., this is not the first time the Leader has sent this request), then the
Helper can either re-compute the current round prepare message or re-transmit a
previously computed message. The Helper MUST verify that the contents of the
`AggregationJob` are identical to the previous message (i.e., containing the
same set of reports).

If the Leader's `round` is behind or more than one round ahead of the Helper's
current round, then the Helper MUST abort with an error of type `roundMismatch`.

The Helper's response to the Leader is an `AggregationJob` (see
{{aggregation-job-representation}}) compiled from the `PrepareStep`s. The order
of the Helper's `prepare_steps` MUST match that of the Leader's.

[[OPEN ISSUE: consider relaxing this ordering constraint. See issue#217.]]

##### DELETE `/tasks/{task-id}/aggregation_jobs/{aggregation-job-id}` {#aggregation-job-delete}

Indicates to the Helper that it MAY abandon the aggregation job and discard all

> state related to it. Whether or not it deletes any data, the Helper MUST respond

Reviewer: I think this text should allow returning e.g. `500` on error -- as written, the Helper must unconditionally respond with a `204`.

Author: - [ ] clean up language about HTTP status code in {{aggregation-job-delete}}

`report_id` field). Each report that was marked as invalid is assigned the
`PrepareStepState` `failed`. Otherwise, the `PrepareStepState` is either
`continued` with the output `prep_msg`, or `finished` if the VDAF preparation
process is finished for the report share.

##### POST `/tasks/{task-id}/aggregation_jobs/{aggregation-job-id}` {#aggregation-job-post}

In the continuation phase, the Leader drives the VDAF preparation of each share
in the candidate report set until the underlying VDAF moves into a terminal
state, yielding an output share for all aggregators or an error. This phase may
involve multiple rounds of interaction depending on the underlying VDAF. Each
round trip is initiated by the Leader making a POST request to an aggregation
job resource.

###### Leader Continuation {#aggregation-leader-continuation}

The Leader begins each round of continuation with a sequence of `PrepareStep`s
received from the Helper and its locally computed prepare message.

If `PrepareStep` is of type `failed`, then the Leader acts based on the value of
the `ReportShareError`:

* If the error is `ReportShareError.report_too_early`, then the Leader MAY try

>   to re-send the report in a later `AggregationJob`.

Author: This behavior is not new. The equivalent passage in DAP-03: https://datatracker.ietf.org/doc/html/draft-ietf-ppm-dap-03#section-4.4.2.1-2.1

##### Representation {#aggregation-job-representation}

~~~
opaque AggregationJobId[16];

struct {
  ReportMetadata metadata;
  opaque public_share<0..2^32-1>;
  HpkeCiphertext encrypted_input_share;
} ReportShare;

struct {
  QueryType query_type;
  select (PartialBatchSelector.query_type) {
    case time_interval: Empty;
    case fixed_size: BatchID batch_id;
  };
} PartialBatchSelector;

enum {
  start(0),
  continued(1),
  finished(2),
  failed(3),
  (255),
} PrepareStepState;

enum {
  batch_collected(0),
  report_replayed(1),
  report_dropped(2),
  hpke_unknown_config_id(3),
  hpke_decrypt_error(4),
  vdaf_prep_error(5),
  batch_saturated(6),
  task_expired(7),
  unrecognized_message(8),
  report_too_early(9),
  (255)
} ReportShareError;

struct {
  ReportID report_id;
  PrepareStepState prepare_step_state;
  select (PrepareStep.prepare_step_state) {
    case start: ReportShare;
    case continued: opaque prep_msg<0..2^32-1>;
    case finished: Empty;
    case failed: ReportShareError;
  }
} PrepareStep;

struct {
  u16 round;
  opaque agg_param<0..2^32-1>;
  PartialBatchSelector part_batch_selector;
  PrepareStep prepare_steps<1..2^32-1>;

> } AggregationJob;

Author: I've been working on implementing this in Janus (https://github.com/divviup/janus/pull/731/files) and I'm finding this definition of `AggregationJob` a little awkward. The idea here is to have a single representation of an aggregation job for both helper and leader across multiple rounds of preparation. But we end up with some extra complexity in implementations:

- Not all `PrepareStep.prepare_step_state` values are valid in all rounds of preparation. For instance, when the leader first sends an `AggregationJob` to the helper, the helper must verify that each `PrepareStep` is in the `start` state. In subsequent rounds, the `start` state is illegal.
- `PrepareStep`s in the `start` state contain a `ReportShare`, which contains `ReportMetadata`, which has `ReportID` in it. That duplicates the `ReportID` field in the outer `PrepareStep` structure, which is wasteful.
- the `agg_param` and `part_batch_selector` fields are only really useful in the first leader->helper messages. Subsequently, the helper has to check that they haven't changed since an earlier round.

All of these mean extra code for servers to write to validate messages. I think we were better off the way things were before, with distinct representations for the initial creation of an aggregation job and then the subsequent stepping. I think it's less REST-y, but that doesn't seem like a good enough tradeoff to me against the implementation complexity and increased message size.

Reviewer 2: +1, I think we should go back to what we had before (more or less). My main concern: We should not transmit the agg param more than once per agg job, since for Poplar this might be relatively large. (Think 1-2MBs.)

The Leader begins an aggregation job by choosing a set of candidate reports that
pertain to the same DAP task and compiling them into an `AggregationJob` (see
{{aggregation-job-representation}}. The Leader can prepare many reports in
parallel by concurrently creating multiple aggregation jobs. Creating
aggregation jobs is how the Leader transmits report shares to the helper, so
each `PrepareStep.prepare_step_state` MUST be set to `PrepareStepState.start`
and the `PrepareStep` MUST contain a `ReportShare`. `AggregationJob.round` MUST
be 0.

If successful, the Helper's responds with a `struct AggregationJob` whose
`round` field is 1 and whose `prepare_steps` is a list of `struct PrepareStep`s
in state `continued` if initialization succeeded or errors if any report was
invalid.

##### POST `/tasks/{task-id}/aggregation_jobs/{aggregation-job-id}` {#aggregation-job-post}

The Leader drives the continuation of input preparation using POST requests to
aggregation job resources (see {{agg-continue-flow}}).

###### DELETE `/tasks/{task-id}/aggregation_jobs/{aggregation-job-id}`

Once an aggregation job is deleted, the Helper MAY abandon it and discard all

> state related to it (e.g., report shares, prepare messages, preparation state).

Reviewer 2: Is the Helper required to keep the agg job state around until it is DELETEd? As a deployer, I don't want to have to count on my Leader to manage my state responsibly.

  definition of Poplar is done.]]

* Information used by the Aggregators to determine how to aggregate each report:

    * For fixed_size tasks, the Leader specifies a "batch ID" that determines
      the batch to which each report for this aggregation job belongs.

      [OPEN ISSUE: For fixed_size tasks, the Leader is in complete control over
      which batch a report is included in. For time_interval tasks, the Client
      has some control, since the timestamp determines which batch window it
      falls in. Is this desirable from a privacy perspective? If not, it might
      be simpler to drop the timestamp altogether and have the agg init request
      specify the batch window instead.]

  The indicated query type MUST match the task's query type. Otherwise, the
  Helper MUST abort with error "queryMismatch".

* The sequence of report shares to aggregate. The `encrypted_input_share` field
  of the report share is the `HpkeCiphertext` whose index in
  `Report.encrypted_input_shares` is equal to the index of the aggregator in the
  task's `aggregator_endpoints` to which the AggregateInitializeReq is being
  sent.

#### Required methods

> ##### PUT `/tasks/{task-id}/aggregation_jobs` {#aggregation-job-put}

Reviewer 2: * Editorial: It might be helpful to define the resource and message format before getting into the protocol flow.
* Where is the aggregation job ID? We still need a way to link requests for the same job across rounds.

Reviewer 2: Summary of offline discussion: The aggregation job ID is missing here because it has been folded into the redirect URI that the Leader is supposed to post to next to move aggregation forward. (In other words, the aggreagtion job ID is now an implementation detail.) The ideas is to make aggreagtion jobs "asynchronous" the way collection jobs are. However there is an important difference between the two: Here the Helper computes its response immediately, i.e., aggregation is inherently synchronous. Collection is inherently asynchronous because the Leader usually won't be finished computing the aggregate at the time the request is made.

We decided that we should re-align this part with DAP-03, i.e., move the aggregation job ID into the API here.

Author: We reached a similar conclusion about the helper's aggregate share resource: the leader should always be able to wait until all relevant aggregation jobs are driven to completion before requesting an aggregate share from the helper. So servicing such a request should just be a matter of collapsing a bunch of output shares into an aggregate share (or, depending on the implementation, collapsing a bunch of partial aggregates over, say, an hour's worth of output shares into the requested aggregate share), which should be doable synchronously for the helper. So I'm also going to move the aggregate share resource back to how it worked in DAP-03 (though we'll still change the request path).

Author: > Editorial: It might be helpful to define the resource and message format before getting into the protocol flow.

I'll give that a try after the dust has settled on other changes.

- [ ] put the description of aggregate job endpoints and messages before the prose discussion of the protocol

The Leader begins an aggregation job by choosing a set of candidate reports that
pertain to the same DAP task and compiling them into an `AggregationJob` (see
{{aggregation-job-representation}}. The Leader can prepare many reports in
parallel by concurrently creating multiple aggregation jobs. Creating
aggregation jobs is how the Leader transmits report shares to the helper, so
each `PrepareStep.prepare_step_state` MUST be set to `PrepareStepState.start`
and the `PrepareStep` MUST contain a `ReportShare`. `AggregationJob.round` MUST
be 0.

If successful, the Helper's responds with HTTP status 303 See Other and a

> Location header containing a URI for the aggregation job. The structure of this

Author: I'm not certain if explicitly spelling out HTTP 303 See Other and a Location header is a good idea. I think we could just say something like "the Helper redirects to a URI for the aggregation job", and rely on existing HTTP specifications for things like the status code and the headers and so on. On the other hand, if we specify this explicitly as I have here, then there's less ambiguity and less code to write for DAP implementations (with this text, they can simply ignore any 3xx status that isn't 303 instead of having to reason about the nuanced difference between 301 and 303). So I lean towards being specific.

## Collecting Results {#collect-flow}

In this phase, the Collector requests aggregate shares from each aggregator and
then locally combines them to yield a single aggregate result. This interaction
is mediated through the Leader, which is responsible for obtaining aggregate
shares from the Helper and relaying them to the Collector. In particular, the
Collector issues a query to the Leader ({{query}}), which the Aggregators use to
select a batch of reports to aggregate. Each emits an aggregate share encrypted
to the Collector so that it can decrypt and combine them to yield the aggregate
result. This entire process revolves around two resources:

1. The Collector uses the collections resource exposed by the Leader (see
   {{collection-resource}}).
1. The Leader uses the aggregate shares resource exposed by the Helper (see
   {{aggregate-share-resource}}).

Once complete, the collector computes the final aggregate result as specified in
{{collect-finalization}}.

### Collection Resource {#collection-resource}

> This resource allows the Collector to create new collections in the Leader and

Reviewer 2: I don't recall if we agreed to be consistent about this, but just in case:
```suggestion
This resource allows the Collector to create new collection jobs in the Leader and
```

Author: I prefer "collections" over "collection jobs". The collection _job_ is the means by which we get the thing we actually care about, which is the collection. But at the end of the day, it doesn't really make a difference which string we use here.

check on their state. Only the Leader supports this resource.

#### Required Methods

> ##### PUT `/tasks/{task-id}/collections` {#collections-put}

Reviewer 2: ```suggestion
##### PUT `/tasks/{task-id}/collection_jobs` {#collections-put}
```

  query type. Otherwise, the Leader MUST abort with error `queryMismatch`.
* `agg_param`, an aggregation parameter for the VDAF being executed. This is the
  same value as in `AggregationJob` (see {{aggregation-job-representation}}).

Depending on the VDAF scheme and how the Leader is configured, the Leader and
Helper may already have prepared a sufficient number of reports satisfying the
query and be ready to return the aggregate shares right away, but this cannot be
guaranteed. In fact, for some VDAFs, it is not be possible to begin preparing
inputs until the collector provides the aggregation parameter in the
`CollectReq`. For these reasons, collect requests are handled asynchronously.

Upon receipt of a `CollectReq`, the Leader begins by checking that the query is
valid using the procedure in {{batch-validation}}. If so, it immediately sends
the Collector a response with HTTP status 303 See Other and a Location header
containing a URI for the individual collection that can be polled by the
Collector, as described in {{collection-resource}}. The structure of this URI is
not specified, so implementations may use any unique identifier they wish for
collections.

> The Leader then asynchronously begins working with the Helper to prepare the
> input shares satisfying the query (or continues this process, depending on the
> VDAF) as described in {{aggregate-flow}}, and then invokes the aggregate share

Reviewer 2: ```suggestion
The Leader then asynchronously begins working with the Helper to aggregate the reports satisfying the query (or continues this process, depending on the
VDAF) as described in {{aggregate-flow}}, and then invokes the aggregate share
```

request flow described in {{aggregate-share-resource}} to gather aggregate
shares from Helpers.

##### POST `{collection URI}`

The collector makes a POST request to the collection resource to check whether
it is available yet. A collection is available if:

> * The Leader has all the output shares it needed to construct its own aggregate
>   share.

Reviewer 2: ```suggestion
* The Leader and Helper have aggregated a sufficient number of reports.
```

* The Leader has obtained the Helper's aggregate share (see
  {{aggregate-share-resource}}).

If the collection is known to the Leader but is not available yet, the Leader
responds with HTTP status 202 Accepted, and MAY include a Retry-After header
to suggest a polling interval to the collector.

If the collection is available, then the Leader computes its own aggregate share
by aggregating all of the prepared output shares that fall within the batch
interval:

~~~
agg_share = VDAF.out_shares_to_agg_share(agg_param, out_shares)
~~~

Where `agg_param` is the VDAF aggregation parameter provided by the Collector
earlier in the `CollectReq`, and `out_shares` is the list of output shares
obtained by executing the aggregation sub-protocol (see {{aggregate-flow}}).

The Leader encrypts `agg_share` under the collector's HPKE public key as
described in {{aggregate-share-encrypt}}.

The Leader then responds to the Collector with a `Collection`:

~~~
struct {
  PartialBatchSelector part_batch_selector;
  uint64 report_count;
  HpkeCiphertext encrypted_agg_shares<1..2^32-1>;

> } Collection;

Reviewer 2: ?
```suggestion
} CollectJobResult;
```

~~~

[OPEN ISSUE: Decide if and how the collector's request is authenticated. If not,
then we need to ensure that collect job URIs are resistant to enumeration
attacks.]

* `query`, the Collector's query. The indicated query type MUST match the task's
  query type. Otherwise, the Leader MUST abort with error `queryMismatch`.
* `agg_param`, an aggregation parameter for the VDAF being executed. This is the
  same value as in `AggregationJob` (see {{aggregation-job-representation}}).

Depending on the VDAF scheme and how the Leader is configured, the Leader and
Helper may already have prepared a sufficient number of reports satisfying the
query and be ready to return the aggregate shares right away, but this cannot be
guaranteed. In fact, for some VDAFs, it is not be possible to begin preparing
inputs until the collector provides the aggregation parameter in the
`CollectReq`. For these reasons, collect requests are handled asynchronously.

Upon receipt of a `CollectReq`, the Leader begins by checking that the query is
valid using the procedure in {{batch-validation}}. If so, it immediately sends
the Collector a response with HTTP status 303 See Other and a Location header

> containing a URI for the individual collect job that can be polled by the

Reviewer: Since we are defining URIs for everything else, is there any reason not to introduce a `CollectJobID` and treat collect jobs as we do e.g. reports or aggregation jobs? (i.e. define a specific URI scheme, including the collect job ID, something like `/tasks/{task-id}/collect_jobs/{collect-job-id}`)

This would match the rest of the API much more closely, and I don't think there are any downsides. The collector would be generating the collect job IDs, but I don't think that is a problem.

Author: Yeah the argument from consistency seems reasonable. The counter-argument is that not specifying the collect job URI is the least change from DAP-03, and thus the least disruptive to implementations. I can go either way on this. I'm curious what @cjpatton thinks.

Reviewer: I hear you; my thoughts are that we should aim to design things as they "should be" if we were writing the design down for the first time, without necessarily worrying about the size of the diff against previous drafts.

Author: OK, that sounds like a consensus.

- [ ] specify a `CollectionJobId` and collection job URIs

* The number of reports included in the batch.

* The vector of Leader and Helper encrypted aggregate shares. They MUST appear
  in the same order as the aggregator endpoints list of the task parameters.

If obtaining Helper aggregate shares fails, then the Leader responds to
subsequent POST requests to the collection with an HTTP error status and a
problem document as described in {{errors}}.

The Leader MAY respond with 204 No Content for requests to a collection if the
results have been deleted due to age.

> ###### A note on idempotence

Reviewer 2: ```suggestion
###### A Note on Idempotence
```

The reason we use a POST instead of a GET to poll the state of a collection is
because of the fixed-size query mode (see {{fixed-size-query}}). Collectors may
make a query against the current batch, and it is the Leader's responsibility to
keep track of what batch is current for some task. Polling a collection is the
only point at which it is safe for the Leader to change its current batch, since
it constitutes acknowledgement on the Collector's part that it received the

> response to some previous PUT request to the collections resource.

Reviewer 2: I don't think I follow. For starters, you're talking about polling the collect URI here, correct? If yes, then the contents of the response, that is the aggregate result, is not going to change once computed.

Author: Yes, but that's not what affects idempotence. What matters is that polling a collection has the _side effect_ of changing what the leader considers to be the current batch. There's a much more verbose explanation of this [here](https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/blob/134d5305849f52a59061a543e843387564b24a69/api-resources.md?plain=1#L341).
